modules/renter/worker.go

```diff
@@ -16,103 +14,160 @@ import (
-	uploadConsecutiveFailures int
-
-	// Two lists of chunks that relate to worker upload tasks. The first list is
-	// the set of chunks that the worker hasn't examined yet. The second list is
-	// the list of chunks that the worker examined, but was unable to process
-	// because other workers had taken on all of the work already. This list is
-	// maintained in case any of the other workers fail - this worker will be
-	// able to pick up the slack.
-	mu                sync.Mutex
-	standbyChunks     []*unfinishedChunk
-	terminated        bool
-	unprocessedChunks []*unfinishedChunk
+	// Download variables.
+	downloadChan                chan struct{}              // Notifications of new work. Takes priority over uploads.
+	downloadChunks              []*unfinishedDownloadChunk // Yet unprocessed work items.
+	downloadConsecutiveFailures int                        // How many failures in a row?
+	downloadRecentFailure       time.Time                  // How recent was the last failure?
+	downloadTerminated          bool                       // Has downloading been terminated for this worker?
+
+	// Upload variables.
+	unprocessedChunks         []*unfinishedChunk // Yet unprocessed work items.
+	uploadChan                chan struct{}      // Notifications of new work.
+	uploadConsecutiveFailures int                // How many times in a row uploading has failed.
+	uploadRecentFailure       time.Time          // How recent was the last failure?
+	uploadTerminated          bool               // Have we stopped uploading?
+
+	// Utilities.
+	//
+	// The mutex is only needed when interacting with 'downloadChunks' and
+	// 'unprocessedChunks', as everything else is only accessed from the single
+	// master thread.
+	killChan chan struct{} // Worker will shut down if a signal is sent down this channel.
+	mu       sync.Mutex
 }
 
-// threadedWorkLoop repeatedly issues work to a worker, stopping when the worker
-// is killed or when the thread group is closed.
-func (w *worker) threadedWorkLoop() {
-	err := w.renter.tg.Add()
-	if err != nil {
+// managedKillDownloading will drop all of the pieces given to the worker.
+func (w *worker) managedKillDownloading() {
+	w.mu.Lock()
+	w.dropDownloadChunks()
+	w.downloadTerminated = true
+	w.mu.Unlock()
+}
+
+// managedKillUploading will disable all uploading for the worker.
+func (w *worker) managedKillUploading() {
+	w.mu.Lock()
+	w.dropUploadChunks()
+	w.uploadTerminated = true
+	w.mu.Unlock()
+}
+
+// onDownloadCooldown returns true if the worker is on cooldown from failed
+// downloads.
+func (w *worker) onDownloadCooldown() bool {
+	requiredCooldown := downloadFailureCooldown
+	for i := 0; i < w.downloadConsecutiveFailures && i < maxConsecutivePenalty; i++ {
+		requiredCooldown *= 2
+	}
+	return time.Now().Before(w.downloadRecentFailure.Add(requiredCooldown))
+}
+
+// onUploadCooldown returns true if the worker is on cooldown from failed
+// uploads.
+func (w *worker) onUploadCooldown() bool {
```

Comment:
Should we separate upload and download cooldowns? Many issues probably affect both.

Reply:
many issues, such as `host has no remaining disk space` only affect uploads.